File: client/src/services/ticketService.js
import axios from 'axios';
import { handleApiError } from '../utils/errorHandler';

const API_URL = import.meta.env.VITE_API_URL;

const ticketService = {
    getTickets: async (orderBy = 'created_at', orderType = 'desc', page = 1, perPage = 20, filters = {}) => {
        try {
            const params = {
                order_by: orderBy,
                order_type: orderType,
                page: page,
                per_page: perPage,
                ...filters
            };

            const response = await axios.get(`${API_URL}/api/tickets`, {
                params: params,
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching tickets:", error);
            throw handleApiError(error);
        }
    },

    getTicket: async (id) => {
        try {
            const response = await axios.get(`${API_URL}/api/tickets/${id}`, { withCredentials: true });
            return response.data;
        } catch (error) {
            console.error("Error fetching ticket:", error);
            throw handleApiError(error);
        }
    },

    getTicketConversations: async (ticketId) => {
        try {
            const response = await axios.get(`${API_URL}/api/tickets/${ticketId}/conversations`, {
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching conversations:", error);
            throw handleApiError(error);
        }
    },

    addReply: async (ticketId, replyData, attachments = []) => {
        try {
            const formData = new FormData();

            Object.entries(replyData).forEach(([key, value]) => {
                if (value !== null && value !== undefined) {
                    formData.append(`reply[${key}]`, value);
                }
            });

            if (attachments && attachments.length > 0) {
                attachments.forEach((file, index) => {
                    formData.append(`attachments[${index}]`, file);
                });
            }

            const response = await axios.post(
                `${API_URL}/api/tickets/${ticketId}/reply`,
                formData,
                {
                    withCredentials: true,
                    headers: {
                        'Content-Type': 'multipart/form-data'
                    }
                }
            );
            return response.data;
        } catch (error) {
            console.error("Error adding reply:", error);
            throw handleApiError(error);
        }
    },

    addNote: async (ticketId, noteData, attachments = []) => {
        try {
            const formData = new FormData();

            Object.entries(noteData).forEach(([key, value]) => {
                if (value !== null && value !== undefined) {
                    if (typeof value === 'boolean') {
                        formData.append(`note[${key}]`, value ? 'true' : 'false');
                    } else {
                        formData.append(`note[${key}]`, value);
                    }
                }
            });

            if (attachments && attachments.length > 0) {
                attachments.forEach((file, index) => {
                    formData.append(`attachments[${index}]`, file);
                });
            }

            const response = await axios.post(
                `${API_URL}/api/tickets/${ticketId}/note`,
                formData,
                {
                    withCredentials: true,
                    headers: {
                        'Content-Type': 'multipart/form-data'
                    }
                }
            );
            return response.data;
        } catch (error) {
            console.error("Error adding note:", error);
            throw handleApiError(error);
        }
    },

    deleteConversation: async (conversationId) => {
        try {
            console.log("Deleting conversation with ID:", conversationId);
            const response = await axios.delete(
                `${API_URL}/api/conversations/${conversationId}`,
                { withCredentials: true }
            );
            return response.data;
        } catch (error) {
            console.error("Error deleting conversation:", error);
            throw handleApiError(error);
        }
    },

    updateConversation: async (conversationId, formData) => {
        try {
            const response = await axios.put(
                `${API_URL}/api/conversations/${conversationId}`,
                formData,
                {
                    withCredentials: true,
                    headers: {
                        'Content-Type': 'multipart/form-data'
                    }
                }
            );
            return response.data;
        } catch (error) {
            console.error("Error updating conversation:", error);
            throw handleApiError(error);
        }
    },

    forwardTicket: async (ticketId, forwardData) => {
        try {
            const payload = {
                body: forwardData.body,
                to_emails: forwardData.to_emails || [],
                cc_emails: forwardData.cc_emails || [],
                bcc_emails: forwardData.bcc_emails || []
            };

            if (forwardData.hasOwnProperty('include_original_attachments')) {
                payload.include_original_attachments = forwardData.include_original_attachments;
            }

            const response = await axios.post(
                `${API_URL}/api/tickets/${ticketId}/forward`,
                payload,
                {
                    withCredentials: true,
                    headers: {
                        'Content-Type': 'application/json'
                    }
                }
            );
            return response.data;
        } catch (error) {
            console.error("Error forwarding ticket:", error);
            throw handleApiError(error);
        }
    },

    createTicket: async (ticketData, attachments = []) => {
        try {
            const formData = new FormData();

            Object.entries(ticketData).forEach(([key, value]) => {
                if (value !== null && value !== undefined && value.length !== 0) {
                    if(key === 'tags' && Array.isArray(value)) {
                        value.forEach((tag) => {
                            formData.append(`ticket[${key}][]`, tag);
                        });
                    } else {
                        formData.append(`ticket[${key}]`, value);
                    }
                }
            });

            if (attachments && attachments.length > 0) {
                attachments.forEach((file, index) => {
                    formData.append(`attachments[${index}]`, file);
                });
            }

            const response = await axios.post(`${API_URL}/api/tickets`, formData, {
                withCredentials: true,
                headers: {
                    'Content-Type': 'multipart/form-data'
                }
            });

            return response.data;
        } catch (error) {
            console.error("Error in createTicket service:", error);
            throw handleApiError(error);
        }
    },

    updateTicket: async (id, ticketData) => {
        try {
            const response = await axios.put(`${API_URL}/api/tickets/${id}`, { ticket: ticketData }, {
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error("Error updating ticket:", error);
            throw handleApiError(error);
        }
    },

    deleteTicket: async (id) => {
        try {
            await axios.delete(`${API_URL}/api/tickets/${id}`, { withCredentials: true });
            return { success: true };
        } catch (error) {
            console.error("Error deleting ticket:", error);
            throw handleApiError(error);
        }
    },

    getTicketFields: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/tickets/fields`, { withCredentials: true });
            return response.data.fields;
        } catch (error) {
            console.error("Error fetching ticket fields:", error);
            throw handleApiError(error);
        }
    },

    getTicketCount: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/tickets/count`, { withCredentials: true });
            return response.data;
        } catch (error) {
            console.error("Error fetching ticket count:", error);
            throw handleApiError(error);
        }
    },

    mergeTickets: async (primaryTicketId, secondaryTicketIds) => {
        try {
            const response = await axios.put(
            `${API_URL}/api/tickets/${primaryTicketId}/merge`,
            { ticket_ids: secondaryTicketIds },
            { withCredentials: true }
            );
            return response.data;
        } catch (error) {
            console.error("Error merging tickets:", error);
            throw handleApiError(error);
        }
    },

    exportTickets: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/tickets/export`, {
                responseType: 'blob',
                withCredentials: true
            });

            const blob = new Blob([response.data], { type: response.headers['content-type'] });
            const url = window.URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = `tickets_${new Date().toISOString()}.csv`;
            a.click();
            window.URL.revokeObjectURL(url);
        } catch (error) {
            console.error("Error exporting tickets:", error);
            throw handleApiError(error);
        }
    }
};

export default ticketService;

File: server/app/services/freshdesk/ticket_service.rb
module Freshdesk
    class TicketService
        def initialize
            @client = ApiClient.new
        end

        def get_ticket(id)
            begin
                @client.request(:get, "/tickets/#{id}")
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Ticket ##{id} not found")
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to get ticket ##{id}: #{e.message}")
            end
        end

        def create_ticket(ticket_params, attachments = [])
            begin
                if attachments.empty?
                    @client.request(:post, '/tickets', ticket_params)
                else
                    @client.request_with_attachments('/tickets', ticket_params, attachments, :post)
                end
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to create ticket: #{e.message}")
            end
        end

        def update_ticket(id, ticket_params)
            begin
                @client.request(:put, "/tickets/#{id}", ticket_params)
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Ticket ##{id} not found")
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to update ticket ##{id}: #{e.message}")
            end
        end

        def delete_ticket(id)
            begin
                @client.request(:delete, "/tickets/#{id}")
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Ticket ##{id} not found")
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to delete ticket ##{id}: #{e.message}")
            end
        end

        def get_ticket_fields
            begin
                @client.request(:get, '/ticket_fields')
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to get ticket fields: #{e.message}")
            end
        end

        def get_conversations(ticket_id)
            begin
                @client.request(:get, "/tickets/#{ticket_id}/conversations")
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Ticket ##{ticket_id} not found")
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to get conversations for ticket ##{ticket_id}: #{e.message}")
            end
        end

        def add_reply(ticket_id, reply_data, attachments = [])
            begin
                if attachments.empty?
                    @client.request(:post, "/tickets/#{ticket_id}/reply", reply_data)
                else
                    @client.request_with_attachments("/tickets/#{ticket_id}/reply", reply_data, attachments, :post)
                end
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Ticket ##{ticket_id} not found")
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to add reply to ticket ##{ticket_id}: #{e.message}")
            end
        end

        def add_note(ticket_id, note_data, attachments = [])
            begin
                if attachments.empty?
                    @client.request(:post, "/tickets/#{ticket_id}/notes", note_data)
                else
                    @client.request_with_attachments("/tickets/#{ticket_id}/notes", note_data, attachments, :post)
                end
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Ticket ##{ticket_id} not found")
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to add note to ticket ##{ticket_id}: #{e.message}")
            end
        end

        def forward_ticket(ticket_id, forward_data)
            begin
                endpoint = "/tickets/#{ticket_id}/forward"
                forward_data[:include_original_attachments] = false unless forward_data.key?(:include_original_attachments)

                @client.request(:post, endpoint, forward_data)
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Ticket ##{ticket_id} not found")
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to forward ticket ##{ticket_id}: #{e.message}")
            end
        end

        def delete_conversation(note_id)
            begin
                @client.request(:delete, "/conversations/#{note_id}")
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Note ##{note_id} not found")
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to delete note ##{note_id}: #{e.message}")
            end
        end

        def update_conversation(note_id, note_data, attachments = [])
            begin
                endpoint = "/conversations/#{note_id}"

                if attachments.empty?
                    @client.request(:put, endpoint, note_data)
                else
                    @client.request_with_attachments(endpoint, note_data, attachments, :put)
                end
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Note ##{note_id} not found")
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to update note ##{note_id}: #{e.message}")
            end
        end

        def merge_tickets(primary_ticket_id, secondary_ticket_ids)
            begin
              primary_ticket_id = primary_ticket_id.to_i

              secondary_ticket_ids = [secondary_ticket_ids].flatten.map(&:to_i)

              endpoint = "/tickets/merge"
              merge_data = { primary_id: primary_ticket_id, ticket_ids: secondary_ticket_ids }

              @client.request(:put, endpoint, merge_data)
            rescue Freshdesk::ResourceNotFoundError
              raise Freshdesk::ResourceNotFoundError.new("Primary ticket ##{primary_ticket_id} not found")
            rescue Freshdesk::ValidationError => e
              raise e
            rescue Freshdesk::Error => e
              raise e
            rescue => e
              raise Freshdesk::RequestError.new("Failed to merge tickets: #{e.message}")
            end
        end
    end
end
